src/codecs/encoding.rs

```diff
@@ -1,2 +1,201 @@
+/// implementation.
+#[typetag::serde(tag = "method")]
+pub trait FramingConfig: Debug + DynClone + Send + Sync {
+    /// Builds a framer from this configuration.
+    ///
+    /// Fails if the configuration is invalid.
+    fn build(&self) -> crate::Result<BoxedFramer>;
+}
+
+dyn_clone::clone_trait_object!(FramingConfig);
+
+/// Serialize a structured event into a byte frame.
+pub trait Serializer:
+    tokio_util::codec::Encoder<Event, Error = crate::Error> + DynClone + Debug + Send + Sync
+{
+}
+
+/// Default implementation for `Serializer`s that implement
+/// `tokio_util::codec::Encoder`.
+impl<Encoder> Serializer for Encoder where
+    Encoder: tokio_util::codec::Encoder<Event, Error = crate::Error> + Clone + Debug + Send + Sync
+{
+}
+
+dyn_clone::clone_trait_object!(Serializer);
+
+/// A `Box` containing a `Serializer`.
+pub type BoxedSerializer = Box<dyn Serializer>;
+
+/// Define options for a serializer and build it from the config object.
+///
+/// Implementors must annotate the struct with `#[typetag::serde(name = "...")]`
+/// to define which value should be read from the `codec` key to select their
+/// implementation.
+#[typetag::serde(tag = "codec")]
+pub trait SerializerConfig: Debug + DynClone + Send + Sync {
+    /// Builds a serializer from this configuration.
+    ///
+    /// Fails if the configuration is invalid.
+    fn build(&self) -> crate::Result<BoxedSerializer>;
+}
+
+dyn_clone::clone_trait_object!(SerializerConfig);
+
+#[derive(Debug, Clone)]
+/// An encoder that can encode structured events into byte frames.
+pub struct Encoder {
+    framer: BoxedFramer,
+    serializer: BoxedSerializer,
+}
+
+impl Default for Encoder {
+    fn default() -> Self {
+        Self {
+            framer: Box::new(NewlineDelimitedEncoder::new()),
+            serializer: Box::new(TextSerializer::new()),
+        }
+    }
+}
+
+impl Encoder {
+    /// Creates a new `Encoder` with the specified `Serializer` to produce bytes
+    /// from a structured event, and the `Framer` to wrap these into a byte
+    /// frame.
+    pub fn new(framer: BoxedFramer, serializer: BoxedSerializer) -> Self {
+        Self { framer, serializer }
+    }
+}
+
+impl tokio_util::codec::Encoder<Event> for Encoder {
+    type Error = Error;
+
+    fn encode(&mut self, item: Event, buffer: &mut BytesMut) -> Result<(), Self::Error> {
+        let mut payload = BytesMut::new();
+
+        // Serialize the event.
+        self.serializer
+            .encode(item, &mut payload)
+            .map_err(|error| {
+                emit!(&EncoderSerializeFailed { error: &error });
+                Error::SerializingError(error)
+            })?;
+
+        // Frame the serialized event.
+        self.framer
+            .encode(payload.freeze(), buffer)
+            .map_err(|error| {
+                emit!(&EncoderFramingFailed { error: &error });
+                Error::FramingError(error)
+            })
```

Comment:
I'm wondering about the performance implications here. As is, we'll be guaranteeing one allocation and one copy per event. That doesn't seem like a blocker necessarily, but I don't want us to end up in a position where we need to redesign things in order to make them more efficient. Do you think there's a path with this design to enabling relatively "direct" serialization (i.e. writing straight to the eventual output, without intermediary writes and copies)?

Reply:
Good point, I think writing to the `buffer` input argument directly should work.

I'll have to examine more close what that means for the error case, though. I'm not sure what the caller currently expects from the provided buffer when encoding fails.

Reply:
Implemented in d796987980d4908aba287776915ba44b7f73d20e.

The `Framer` now uses `tokio_util::codec::Encoder<()>` to modify the frame in-place.

Also noteworthy that `FramedWrite` does not always pass an empty buffer to `encode`. I handled that case specifically in 9aa870b01ad308ffe32d81d2823781ac0fda1f89 by making sure the `Framers`/`Serializers` get a view to the buffer that is intended for the payload only.